Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -298,157 +321,309 @@ void GCMemcardManager::SetSlotFileInteractive(int slot)
     }
+    lookup[index] = true;
+  }
 
-    // TODO: This is obviously intended to check for success instead.
-    const auto exportRetval = memcard->ExportGci(file_index, path.toStdString(), "");
-    if (exportRetval == Memcard::GCMemcardExportFileRetVal::UNUSED)
-    {
-      File::Delete(path.toStdString());
-    }
+  std::vector<u8> selected_indices;
+  for (u8 i = 0; i < Memcard::DIRLEN; ++i)
+  {
+    if (lookup[i])
+      selected_indices.push_back(i);
   }
 
-  QString text = count == 1 ? tr("Successfully exported the save file.") :
-                              tr("Successfully exported the %1 save files.").arg(count);
-  ModalMessageBox::information(this, tr("Success"), text);
+  return selected_indices;
 }
 
-void GCMemcardManager::ExportAllFiles()
+static QString GetFormatDescription(Memcard::SavefileFormat format)
 {
-  // This is nothing but a thin wrapper around ExportFiles()
-  m_slot_table[m_active_slot]->selectAll();
-  ExportFiles(false);
+  switch (format)
+  {
+  case Memcard::SavefileFormat::GCI:
+    return QObject::tr("Native GCI File");
+  case Memcard::SavefileFormat::GCS:
+    return QObject::tr("MadCatz Gameshark files");
+  case Memcard::SavefileFormat::SAV:
+    return QObject::tr("Datel MaxDrive/Pro files");
+  default:
+    assert(0);
+    return QObject::tr("Native GCI File");
+  }
 }
 
-void GCMemcardManager::ImportFile()
+void GCMemcardManager::ExportFiles(Memcard::SavefileFormat format)
 {
-  QString path = QFileDialog::getOpenFileName(
-      this, tr("Import Save File"), QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)),
-      tr("Native GCI File (*.gci)") + QStringLiteral(";;") + tr("MadCatz Gameshark files(*.gcs)") +
-          QStringLiteral(";;") + tr("Datel MaxDrive/Pro files(*.sav)"));
+  const auto& memcard = m_slot_memcard[m_active_slot];
+  if (!memcard)
+    return;
 
-  if (path.isEmpty())
+  const auto selected_indices = GetSelectedFileIndices();
+  if (selected_indices.empty())
     return;
 
-  const auto result = m_slot_memcard[m_active_slot]->ImportGci(path.toStdString());
+  const auto savefiles = Memcard::GetSavefiles(*memcard, selected_indices);
+  if (savefiles.empty())
+  {
+    ModalMessageBox::warning(this, tr("Export Failed"),
+                             tr("Failed to read selected savefile(s) from memory card."));
+    return;
+  }
+
+  std::string extension = Memcard::GetDefaultExtension(format);
 
-  if (result != Memcard::GCMemcardImportFileRetVal::SUCCESS)
+  if (savefiles.size() == 1)
   {
-    ModalMessageBox::critical(this, tr("Import failed"), tr("Failed to import \"%1\".").arg(path));
+    // when exporting a single save file, let user specify exact path
+    const std::string basename = Memcard::GenerateFilename(savefiles[0].dir_entry);
+    const QString qformatdesc = GetFormatDescription(format);
+    const QString qfilename = QFileDialog::getSaveFileName(
+        this, tr("Export Save File"),
+        QString::fromStdString(File::GetUserPath(D_GCUSER_IDX)) +
```

Comment:
Perhaps this should be placred into a fmt::format call and then be converted to a QString at the end or something? This formatting has a lot going on churning QStrings from QStrings, and it's pretty wonky to follow.

Reply:
Yeah the string concat is a bit messy in places... I'll see what I can do to make it nicer.

Reply:
Alright, I think this is better now?